compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -1977,8 +1977,9 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             Rvalue::Cast(cast_kind, op, ty) => {
                 self.check_operand(op, location);
 
-                match cast_kind {
-                    CastKind::PointerCoercion(PointerCoercion::ReifyFnPointer) => {
+                match *cast_kind {
+                    CastKind::PointerCoercion(PointerCoercion::ReifyFnPointer, coercion_source) => {
+                        let is_implicit_coercion = coercion_source == CoercionSource::Implicit;
```

Comment:
nitpick: Maybe just have a `CoercionSource` field in `ConstraintCategory::Cast` too, instead of a `bool` and this conversion?

Reply:
I considered this, but decided against it, because the `Ord` of `ConstraintCategory` is relevant for the diagnostic output and I didn't want this property to spread all over the compiler.